resources/type-system/inference.md

```diff
@@ -1324,51 +1360,100 @@ determined as follows:
 
   - Otherwise, let `K_1` be `FutureOr<K>`.
 
-- Let `m_1` be the result of performing expression type inference on `e_1`, in
+- Let `m_1` be the result of performing expression inference on `e_1`, in
   context `K_1`.
 
 - Let `T_1` be the static type of `m_1`.
 
-- Let `T` be `flatten(T_1)`.
-
-- Let `m` be an expression artifact whose runtime behavior is as follows:
-
-  - Execute expression artifact `m_1`, and let `o_1` be the resulting value. _By
-    expression soundness, `o_1` will be an instance of the type `T_1`._
-
-  - Define `o_2` as follows:
-
-    - If `o_1` is a subtype of `Future<T>`, then let `o_2` be `o_1`.
-
-    - Otherwise, let `o_2` be the result of creating a new object using the
-      constructor `Future<T>.value()` with `o_1` as its
-      argument.
-
-      - _TODO(paulberry): How do we ensure that the value passed to
-        `Future<T>.value()` is an instance of type `T`? What we have is that
-        `o_1` is an instance of `T_1`, where `T` is `flatten(T_1)`. Which is not
-        quite the same._
-
-    - _By expression soundness, `o_2` must be an instance of `Future<T>`._
-
-  - Pause the stream associated with the innermost enclosing asynchronous
-    **for** loop, if any.
-
-  - Suspend the current invocation of the function body immediately enclosing
-    the await expression until after `o_2` completes.
-
-  - At some time after `o_2` is completed, control returns to the current
-    invocation.
-
-  - If `o_2` is completed with an error `x` and stack trace `t`, then `m`
-    completes with the exception `x` and stack trace `t`. _This trivially
-    satisfies expression soundess, since expression soundness is only concerned
-    with the value of the expression when it completes normally._
-
-  - If `o_2` is completed with an object `v`, then `m` completes with the value
-    `v`. _Since `o_2` is an instance of `Future<T>`, by future soundness, `v`
-    must be an instance of `T`. This satisfies expression soundness, since `v`
-    is an instance of `T` , and `T` is the static type._
+- If `T_1` is incompatible with await (as defined in the _extension types_
+  specification), then there is a compile-time error.
+
+- Let `T_2` be `flatten(T_1)`.
+
+- Let `m_2` be `@LET(T_1 v = m_1 in v is Future<T_2> ? v :
+  Future<T_2>.value(@PROMOTED_TYPE<T_2>(v)))`.
```

Comment:
Here too I'd introduce a `@FUTURE_VALUE<T_2>(v)`, not just for the usual reasons, but also because I don't want the specification to say which exact constructor must be used, as long as it creates a future of the correct type.

Reply:
I think it’s also important that it creates a future with the correct semantics. Specifying this as `Future<T_2>.value(@PROMOTED_TYPE<T_2>(v))` captures both the static type we want *and* the runtime semantics we want, so I think it would be a high bar to come up with something better. As to whether we want to specify the exact constructor that must be used, I think it goes without saying that the implementation is always welcome to replace one elaborated expression with another one that has equivalent semantics (that’s what optimization is all about, after all). I don't think we need to invent a new piece of syntax here to grant extra permission for optimization in this specific case.

Reply:
I vaguely remember someone filing an issue about some backend *not* using `Future.value` to create the future (maybe they use `_Future.immediateValue` instead), because there was a timing difference between the two.

I'd be *much* happier if we specified this abstractly as introducing an object implementing `Future<T_2>` which completes with the value `v` at some later point. Saying that it *has to* call a specific constructor, one which has *extra overhead* because it can also accept a `Future<T2>` or `null` as argument, and has to check for those, even though we know we'll only call with a `T2` instance.

Can we at least say that it calls `Future<T_2>._value(v)` instead, which is a (hypothetical) constructor with signature `Future<T2> Function(T2 value)` that creates a future completed with `v`?
(When nobody can find that constructor, we'll just say it's platform specific.)

Reply:
Ok, looking at the implementation again, the CFE doesn't actually insert an explicit call to `Future.value`. It simply sets a flag which is documented as telling the back-end to insert an explicit call to `Future.value`.

Silly me for trusting the documentation comment, though, because digging further into the implementations, they actually don't all do precisely the same thing:
- In dart2js, a call is made to a private method called `_wrapAwaitedExpression` (in `sdk/lib/_internal/js_runtime/lib/async_patch.dart`), which calls `_Future.value`.
- In dev_compiler, a call is made to a private method called `_awaitWithTypeCheck` (in `sdk/lib/_internal/js_dev_runtime/patch/async_patch.dart`), which also calls calls `_Future.value`.
- In wasm, a call is made to a private method called `_awaitHelperWithTypeCheck` (in `sdk/lib/_internal/wasm/lib/async_patch.dart`), which is implemented in terms of lower level primitives like `scheduleMicrotask`.
- In the vm, the necessary logic is inlined directly into the method using the lower level primitive `AwaitWithTypeCheck`.

I've reworked this section. Rather than refer to `Future<T_2>.value` or create a new primitive `@FUTURE_VALUE`, I'm introducing a new primitive called `@AWAIT_WITH_TYPE_CHECK`, which encapsulates the behavior of `_wrapAwaitedExpression`, `_awaitWithTypeCheck`, etc.